In [1]:
import pandas as pd
import numpy as np
import matplotlib as mp
import statsmodels.api as sm

from scipy.linalg import toeplitz
from statsmodels.tsa.stattools import acf

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/xiaowanzisj/Data-Science-1/refs/heads/main/dataset_lm.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

In [4]:
df

Dependent Var  Explanatory Var #1  Explanatory Var #2  \
0        56.293458           13.698667           50.639873   
1        58.473431            2.714725           65.845845   
2        94.195330           11.618072           65.072497   
3        29.074583            0.818623           45.408996   
4        86.035569            9.077544           73.548021   
..             ...                 ...                 ...   
417      61.300432            0.338441           70.430598   
418      26.309237           -1.729712           47.087996   
419      58.350627           18.322301           53.267835   
420      31.954003            0.436357           61.844132   
421      50.590958           -3.107425           64.590632   

     Explanatory Var #3  Explanatory Var #4  Explanatory Var #5  \
0                     0          -18.568035           45.121911   
1                     1          -25.105932           47.190213   
2                     0           -7.897464           52.163036   
3                     1          -18.316132           54.356714   
4                     0          -19.204165           47.186807   
..                  ...                 ...                 ...   
417                   0          -21.525979           50.741115   
418                   1          -19.034807           55.242929   
419                   0          -26.186201           36.702958   
420                   0          -25.881961           57.106851   
421                   1          -20.303215           56.374124   

     Explanatory Var #6  Explanatory Var #7  Explanatory Var #8  \
0             11.412501           56.410757                   2   
1             10.080280           65.383107                   3   
2             11.057301           82.812717                   0   
3              5.029029           48.812471                   1   
4             12.128134           62.520911                   2   
..                  ...                 ...                 ...   
417           -8.050843           39.075397                   3   
418           28.001769           76.429649                   0   
419           14.530116           51.275150                   2   
420           21.786066           68.928442                   0   
421            4.604237           49.501005                   0   

     Explanatory Var #9  Explanatory Var #10  Explanatory Var #11  \
0            -12.281132            38.996909            -3.010548   
1            -36.763585            51.654939             4.991111   
2            -15.733547            48.913837            -2.457696   
3            -12.825591            45.851732            14.974177   
4            -13.804860            47.765904             9.593982   
..                  ...                  ...                  ...   
417          -29.197173            57.473862            15.505202   
418          -21.941401            54.641620            14.295688   
419          -20.699670            46.443687             5.708762   
420          -23.166461            59.194840             8.365762   
421          -18.792745            45.594612            16.414449   

     Explanatory Var #12  Explanatory Var #13  Explanatory Var #14  \
0              49.195073                    0           -21.153143   
1              45.591729                    0            -6.474403   
2              56.608806                    0           -27.903299   
3              47.362594                    1           -10.064411   
4              53.700562                    0           -17.546302   
..                   ...                  ...                  ...   
417            73.280605                    1            -3.602482   
418            49.816798                    1           -13.910494   
419            52.751016                    1            -6.202345   
420            64.291056                    1           -21.330017   
421            66.957547                    0      

In [5]:
df.isna().sum()

Dependent Var          0
Explanatory Var #1     0
Explanatory Var #2     0
Explanatory Var #3     0
Explanatory Var #4     0
Explanatory Var #5     0
Explanatory Var #6     0
Explanatory Var #7     0
Explanatory Var #8     0
Explanatory Var #9     0
Explanatory Var #10    0
Explanatory Var #11    0
Explanatory Var #12    0
Explanatory Var #13    0
Explanatory Var #14    0
Explanatory Var #15    0
dtype: int64

### Part A (25%)
1.	Download the “dataset_lm.csv” file from Canvas and upload it to Jupyter Notebook. 
2.	Run the OLS model by using the dependent and explanatory variables in the dataset. 
3.	Show your summary table in Python and interpret your results in the summary report.


In [6]:
y = df['Dependent Var']
x = df.drop('Dependent Var', axis=1)

In [7]:
model_updated = sm.OLS(y,x).fit()
model_updated.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Dependent Var   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                          4.509e+04
Date:                Fri, 18 Oct 2024   Prob (F-statistic):                        0.00
Time:                        21:25:30   Log-Likelihood:                         -841.76
No. Observations:                 422   AIC:                                      1714.
Df Residuals:                     407   BIC:                                      1774.
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
Explanatory Var #1      1.3078      0.013    102.159      0.000       1.283       1.333
Explanatory Var #2      1.7667      0.009    203.101      0.000       1.750       1.784
Explanatory Var #3      6.6083      0.177     37.232      0.000       6.259       6.957
Explanatory Var #4      2.0725      0.011    189.896      0.000       2.051       2.094
Explanatory Var #5     -0.7776      0.011    -68.372      0.000      -0.800      -0.755
Explanatory Var #6      0.0124      0.008      1.595      0.111      -0.003       0.028
Explanatory Var #7      0.0286      0.006      4.890      0.000       0.017       0.040
Explanatory Var #8      0.3510      0.080      4.398      0.000       0.194       0.508
Explanatory Var #9     -0.0293      0.010     -2.838      0.005      -0.050      -0.009
Explanatory Var #10     0.1292      0.011     11.409      0.000       0.107       0.152
Explanatory Var #11     0.0145      0.013      1.152      0.250      -0.010       0.039
Explanatory Var #12     0.0752      0.009      8.698      0.000       0.058       0.092
Explanatory Var #13     0.7414      0.177      4.198      0.000       0.394       1.089
Explanatory Var #14    -0.0218      0.011     -1.974      0.049      -0.043   -8.61e-05
Explanatory Var #15     0.1212      0.012     10.324      0.000       0.098       0.144
==============================================================================
Omnibus:                        0.087   Durbin-Watson:                   1.875
Prob(Omnibus):                  0.957   Jarque-Bera (JB):                0.188
Skew:                          -0.004   Prob(JB):                        0.910
Kurtosis:                       2.897   Cond. No.                         287.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [8]:
x_updated = sm.add_constant(x)
model_updated = sm.OLS(y,x_updated).fit()
model_updated.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 4.504e+30
Date:                Fri, 18 Oct 2024   Prob (F-statistic):               0.00
Time:                        21:25:30   Log-Likelihood:                 12229.
No. Observations:                 422   AIC:                        -2.443e+04
Df Residuals:                     406   BIC:                        -2.436e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  32.0000   5.61e-14    5.7e+14      0.000      32.000      32.000
Explanatory Var #1      1.3000   4.53e-16   2.87e+15      0.000       1.300       1.300
Explanatory Var #2      1.7000   3.29e-16   5.16e+15      0.000       1.700       1.700
Explanatory Var #3      6.2000   6.32e-15   9.81e+14      0.000       6.200       6.200
Explanatory Var #4      2.1000   3.89e-16   5.39e+15      0.000       2.100       2.100
Explanatory Var #5     -0.9000   4.56e-16  -1.97e+15      0.000      -0.900      -0.900
Explanatory Var #6  -1.522e-15   2.75e-16     -5.529      0.000   -2.06e-15   -9.81e-16
Explanatory Var #7  -9.827e-16   2.13e-16     -4.620      0.000    -1.4e-15   -5.65e-16
Explanatory Var #8   -3.22e-15   2.89e-15     -1.114      0.266    -8.9e-15    2.46e-15
Explanatory Var #9  -1.246e-15    3.7e-16     -3.370      0.001   -1.97e-15   -5.19e-16
Explanatory Var #10 -2.045e-15   4.61e-16     -4.440      0.000   -2.95e-15   -1.14e-15
Explanatory Var #11 -1.624e-15   4.45e-16     -3.645      0.000    -2.5e-15   -7.48e-16
Explanatory Var #12   7.98e-17   3.33e-16      0.239      0.811   -5.75e-16    7.35e-16
Explanatory Var #13 -1.657e-14   6.38e-15     -2.595      0.010   -2.91e-14   -4.02e-15
Explanatory Var #14  -7.06e-16   3.92e-16     -1.801      0.072   -1.48e-15    6.45e-17
Explanatory Var #15 -8.431e-16   4.67e-16     -1.806      0.072   -1.76e-15    7.45e-17
==============================================================================
Omnibus:                        0.824   Durbin-Watson:                   1.123
Prob(Omnibus):                  0.662   Jarque-Bera (JB):                0.928
Skew:                          -0.070   Prob(JB):                        0.629
Kurtosis:                       2.819   Cond. No.                     2.55e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.55e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

$$ y = 32.0000 + 1.3000 \cdot X_1 + 1.7000 \cdot X_2 + 6.2000 \cdot X_3 + \dots + (-8.4316 \cdot 10^{-16}) \cdot X_{15} + \epsilon $$

### Part B

In [9]:
# Q1
residuals = model_updated.resid
std_deviation = residuals.std()
std_deviation

4.9602203603458624e-14

In [10]:
acf_residuals = acf(residuals, nlags = 3)
acf_residuals

array([1.        , 0.094231  , 0.05348875, 0.03163002])

In [11]:
# Q2
rho = acf_residuals[1]
n = len(y)
cov_matrix = std_deviation**2 * toeplitz(np.append([1, rho], np.zeros(n-2)))
sm.GLS(y, x_updated, sigma=cov_matrix).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            GLS Regression Results                            
==============================================================================
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            GLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 1.340e+29
Date:                Fri, 18 Oct 2024   Prob (F-statistic):               0.00
Time:                        21:25:30   Log-Likelihood:                 11487.
No. Observations:                 422   AIC:                        -2.294e+04
Df Residuals:                     406   BIC:                        -2.288e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  32.0000   3.26e-13   9.81e+13      0.000      32.000      32.000
Explanatory Var #1      1.3000   2.63e-15   4.95e+14      0.000       1.300       1.300
Explanatory Var #2      1.7000   1.92e-15   8.87e+14      0.000       1.700       1.700
Explanatory Var #3      6.2000   3.65e-14    1.7e+14      0.000       6.200       6.200
Explanatory Var #4      2.1000   2.26e-15   9.29e+14      0.000       2.100       2.100
Explanatory Var #5     -0.9000   2.66e-15  -3.38e+14      0.000      -0.900      -0.900
Explanatory Var #6   -3.79e-16    1.6e-15     -0.238      0.812   -3.52e-15    2.76e-15
Explanatory Var #7  -4.163e-16   1.23e-15     -0.337      0.736   -2.84e-15    2.01e-15
Explanatory Var #8  -2.831e-15   1.68e-14     -0.169      0.866   -3.58e-14    3.01e-14
Explanatory Var #9   1.141e-15   2.14e-15      0.534      0.594   -3.06e-15    5.35e-15
Explanatory Var #10  7.078e-16   2.65e-15      0.267      0.790   -4.51e-15    5.92e-15
Explanatory Var #11 -4.129e-16   2.58e-15     -0.160      0.873   -5.48e-15    4.66e-15
Explanatory Var #12 -3.525e-15   1.92e-15     -1.833      0.068   -7.31e-15    2.56e-16
Explanatory Var #13  2.065e-14   3.67e-14      0.563      0.573   -5.14e-14    9.27e-14
Explanatory Var #14 -1.449e-15   2.27e-15     -0.638      0.524   -5.92e-15    3.02e-15
Explanatory Var #15 -3.327e-15   2.68e-15     -1.242      0.215    -8.6e-15    1.94e-15
==============================================================================
Omnibus:                        0.021   Durbin-Watson:                   0.048
Prob(Omnibus):                  0.989   Jarque-Bera (JB):                0.077
Skew:                          -0.013   Prob(JB):                        0.962
Kurtosis:                       2.939   Cond. No.                     2.33e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.33e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Part C (25%)
1.	Split the dataset into two as the training and test sets (test size = 0.5). 
2.	Run the Lasso model with alpha=1 and estimate the coefficients using the training set. 
3.	Then, calculate the mean absolute percentage error using the test set. 
4.	Find an approximate value for alpha that minimizes the mean absolute percentage error.

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_absolute_percentage_error

In [13]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.5, random_state=42)

In [14]:
lasso = Lasso(alpha=1)
lasso.fit(x_train, y_train)
lasso.coef_

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

# 以下GPT

In [15]:
y_pred = lasso.predict(x_test)
mape = mean_absolute_percentage_error(y_test, y_pred)
mape

0.04432190198291579

In [16]:
alphas = np.logspace(-3, 2, 100)  # Try alpha values between 0.001 and 100
best_alpha = 1
best_mape = float('inf')

for alpha in alphas:
    lasso = Lasso(alpha=alpha)
    lasso.fit(x_train, y_train)
    y_pred = lasso.predict(x_test)
    mape = mean_absolute_percentage_error(y_test, y_pred)
    
    if mape < best_mape:
        best_mape = mape
        best_alpha = alpha

print(f"Best alpha: {best_alpha}")
print(f"Lowest MAPE: {best_mape}")

Best alpha: 0.001
Lowest MAPE: 4.4128318074212444e-05


In [17]:
import numpy as np
from sklearn.metrics import mean_squared_error

# Step 1: Fit the Lasso model
lasso = Lasso(alpha=1)
lasso.fit(x_train, y_train)

# Step 2: Predict values for the training set
y_train_pred = lasso.predict(x_train)

# Step 3: Calculate R-squared
r_squared = lasso.score(x_train, y_train)
print(f"R-squared: {r_squared}")

# Step 4: Calculate Mean Squared Error (MSE) for AIC and BIC
mse = mean_squared_error(y_train, y_train_pred)
n = x_train.shape[0]  # number of observations
k = x_train.shape[1]  # number of features (including the intercept if added)

# Step 5: Calculate AIC
aic = n * np.log(mse) + 2 * k
print(f"AIC: {aic}")

# Step 6: Calculate BIC
bic = n * np.log(mse) + np.log(n) * k
print(f"BIC: {bic}")

# Step 7: Calculate F-statistic (you may need to derive this based on variance analysis)
# There isn't a direct formula for F-statistic in Lasso from sklearn,
# but it's typically used to compare two models. 

R-squared: 0.9940207931005163
AIC: 331.07259183490703
BIC: 381.35046383704804


### Part D (25%)
1.	Use the demand data given in the table and develop an appropriate forecasting model (i.e., the tailored regularization discussed in the class—see your slides for more info) that exploits the available information given in the table as much as possible.
2.	Interpret your results.


In [18]:
data = {
    'Month': list(range(1, 26)),
    'Demand': [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144],
    'Advance_demand': [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]
}

df_d = pd.DataFrame(data)

In [19]:
df_d

Month  Demand  Advance_demand
0       1     100              71
1       2     112              30
2       3     107              75
3       4     103              64
4       5      91              41
5       6      85              51
6       7      84              42
7       8      85              51
8       9      79              57
9      10      81              49
10     11     134             134
11     12      86              52
12     13      99              99
13     14      89              56
14     15     111              81
15     16     114              79
16     17     118              73
17     18     163             163
18     19     193             193
19     20     143              99
20     21     144              91
21     22     202             202
22     23     158             105
23     24     160             101
24     25     144              96

# GPT code

In [20]:
from sklearn.metrics import mean_absolute_error
# Step 1: 准备数据
X = df_d[['Advance_demand']]  # 自变量 (特征)
y = df_d['Demand']  # 因变量 (目标)

# Step 2: 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Step 3: 运行 Lasso 模型
lasso = Lasso(alpha=1)  # alpha 是正则化强度的参数
lasso.fit(X_train, y_train)

# Step 4: 预测和评估模型
y_pred = lasso.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
print(f"Mean Absolute Error: {mae}")

# 打印模型系数
print(f"Model Coefficients: {lasso.coef_}")

Mean Absolute Error: 16.12293300907556
Model Coefficients: [0.69681898]


模型公式：

$ y = \beta_0 + \beta_1 \cdot X_1 + \epsilon $

具体参数代入：

$ y = 0 + 0.6968 \cdot \text{Advance demand} + \epsilon $